feat: make go work with updated FFI

Iandenh opened the pull request:
<!-- Thanks for creating a PR! To make it easier for reviewers and everyone else to understand what your changes relate to, please add some relevant content to the headings below. Feel free to ignore or delete sections that you don't think are relevant. Thank you! ❤️ -->

## About the changes
<!-- Describe the changes introduced. What are they and why are they being introduced? Feel free to also add screenshots or steps to view the changes if they're visual. -->
This updates the FFI implementation of the Go code. This was outdated and was no longer working. I updated the minimum golang to 1.18 in order to use generics.

<!-- (For internal contributors): Does it relate to an issue on public roadmap? -->
<!--
Relates to [roadmap](https://github.com/orgs/Unleash/projects/10) item: #
-->

### Important files
<!-- PRs can contain a lot of changes, but not all changes are equally important. Where should a reviewer start looking to get an overview of the changes? Are any files particularly important?  -->


## Discussion points
- I simply use an empty object for the custom strategy (this is also now deprecated in Unleash, right?). This was also done for `getVariant` in Java
- This part (in tests) I copied from a Java test case:
```golang
 if variant == nil {
    variant = &VariantDef{"disabled", nil, false, engine.IsEnabled("Feature.A", context)}
}
```

sighphyre commented on go-engine/unleash_engine_test.go:
```diff
@@ -212,6 +213,10 @@ func TestClientSpecification(t *testing.T) {
 				result := engine.GetVariant(toggleName, context)
 
+				if result == nil {
+					result = &VariantDef{"disabled", nil, false, engine.IsEnabled("Feature.A", context)}
```
I think this should just be false. Tests pass (on my machine!) with this change but if there's a reason we need to do the evaluation for the actual tests, then probably we should do something like check the actual test toggle that we're testing

```suggestion
					result = &VariantDef{"disabled", nil, false, false}
```

sighphyre commented on go-engine/unleash_engine_test.go:
```diff
@@ -86,8 +86,9 @@ func TestGetVariant(t *testing.T) {
+
 	if variant == nil {
-		t.Fatalf("Feature.A should have variant")
+		variant = &VariantDef{"disabled", nil, false, engine.IsEnabled("Feature.A", context)}
```
Think it's okay to have the actual toggle checked here

sighphyre reviewed (changes requested):
Oh cheers, this looks good to me. I think we should patch the test, as per my suggestion but I think this is great, thank you!

sighphyre commented:
Cheers, thank you so much for tackling this! Just a heads up, the FFI layers for Yggdrasil don't really run anywhere in prod at the moment. We're moving towards swapping out the code of the relevant SDKs with this in the nearish future but that's not done yet. You're more than welcome to use this if you find it useful, the core evaluation logic is very battle tested through the Edge server but the FFI layers are still a work in progress. I'd be super interested to know what you're using it for, if you are!

>     variant = &VariantDef{"disabled", nil, false, engine.IsEnabled("Feature.A", context)}
I think the Java tests might be cheating here, haha! Thanks for raising this, I'll go patch the Java side sometime in the nearish future, I've given you a suggestion for something slightly better in lieu of taking the Java version!

> I simply use an empty object for the custom strategy (this is also now deprecated in Unleash, right?).
I think this is totally fine for this PR. Custom strategies are deprecated but for us (at least on the SDK layer) that means supporting them for a long time yet. They're not that hard to add, the Ruby and C# versions have code to handle this if you're interested but I think the no-op you've put there is perfect for bringing the Go version back to a happy state for now

Iandenh commented:
Hi, thanks for your review and feedback. I have updated the test.

And yes, I'm using this project in a personal project. And with some modifications, it works great. Here is a bit more of a debt story about it: :smile: 

At my job, we recently moved from Overmizly to Unleash due to forcing us to migrate to another plan. When using Overmizly I was a bit unhappy with the poor web app usability and performance. To ensure I didn't need to go to that service as often, I created a small service to sit between the Overmizly CDN and my local development environment. This service can quickly toggle feature flag statuses between enabled and disabled, allowing me to test both flows efficiently.

Despite Unleash being a significant improvement in speed and usability, I refactored my application to support Unleash and act as a sort of proxy with the applied overrides (also to learn a bit about the Unleash APIs). Now, I use it as a companion tool for development. It allows me to quickly toggle feature flag statuses that only affect my environment, preventing disruptions for other developers. The tool still focuses on overriding feature flags and includes some quality-of-life features like fuzzy find and keyboard shortcuts.

I have customized this project to support a subset of the frontend API (supporting a subset of the Client API was much simpler). I found inspiration through reading Unleash Edge documentation. On my fork of this project, I added two extra FFI functions, `resolve` and `resolve_all`, which I discovered by studying the Unleash Edge frontend API implementation. If you are interested in extending the FFI functions of this library, I am more than happy to polish my code and create pull requests for them.

sighphyre reviewed (approved):
LGTM! Thank you again!
